Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp

```diff
@@ -0,0 +1,269 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DolphinQt2/Config/Mapping/IOWindow.h"
+
+#include <thread>
+
+#include <QComboBox>
+#include <QDialogButtonBox>
+#include <QHBoxLayout>
+#include <QLabel>
+#include <QListWidget>
+#include <QPlainTextEdit>
+#include <QPushButton>
+#include <QSlider>
+#include <QSpinBox>
+#include <QVBoxLayout>
+
+#include "Core/Core.h"
+#include "DolphinQt2/Config/Mapping/MappingCommon.h"
+#include "DolphinQt2/Config/Mapping/MappingWidget.h"
+#include "DolphinQt2/Config/Mapping/MappingWindow.h"
+#include "InputCommon/ControlReference/ControlReference.h"
+#include "InputCommon/ControllerEmu/ControllerEmu.h"
+#include "InputCommon/ControllerInterface/ControllerInterface.h"
+
+constexpr int SLIDER_TICK_COUNT = 100;
+
+IOWindow::IOWindow(MappingWidget* parent, ControlReference* ref, IOWindow::Type type)
+    : QDialog(static_cast<QWidget*>(parent)), m_reference(ref), m_parent(parent), m_type(type)
```

Comment:
Since this blocks the parent window while in use and is never reused, there's no need to call functions/get data from the parent. Any needed data should be passed in explicitly.